starknet/src/space/space.cairo

```diff
@@ -490,63 +582,126 @@ mod Space {
+
+                self
+                    .emit(
+                        Event::MaxVotingDurationUpdated(
+                            MaxVotingDurationUpdated {
+                                max_voting_duration: input.max_voting_duration
+                            }
+                        )
+                    );
             }
 
             if NoUpdateU32::should_update(@input.min_voting_duration) {
                 _set_min_voting_duration(ref self, input.min_voting_duration);
-                MinVotingDurationUpdated(input.min_voting_duration);
+
+                self
+                    .emit(
+                        Event::MinVotingDurationUpdated(
+                            MinVotingDurationUpdated {
+                                min_voting_duration: input.min_voting_duration
+                            }
+                        )
+                    );
             }
 
             if NoUpdateU32::should_update(@input.voting_delay) {
                 _set_voting_delay(ref self, input.voting_delay);
-                VotingDelayUpdated(input.voting_delay);
+
+                self
+                    .emit(
+                        Event::VotingDelayUpdated(
+                            VotingDelayUpdated { voting_delay: input.voting_delay }
+                        )
+                    );
             }
 
             if NoUpdateArray::should_update((@input).metadata_URI) {
-                MetadataURIUpdated(@input.metadata_URI);
+                self
+                    .emit(
+                        Event::MetadataURIUpdated(
+                            MetadataURIUpdated { metadata_URI: input.metadata_URI.span() }
+                        )
+                    );
             }
 
             if NoUpdateArray::should_update((@input).dao_URI) {
-                DaoURIUpdated(@input.dao_URI);
+                self.emit(Event::DAOURIUpdated(DAOURIUpdated { dao_URI: input.dao_URI.span() }));
             }
 
             // if not NO_UPDATE
             if NoUpdateStrategy::should_update((@input).proposal_validation_strategy) {
-                // TODO: might be possible to remove need to clone by defining the event or setter on a snapshot.
-                // Similarly for all non value types.
                 _set_proposal_validation_strategy(
                     ref self, input.proposal_validation_strategy.clone()
                 );
-                ProposalValidationStrategyUpdated(
-                    @input.proposal_validation_strategy,
-                    @input.proposal_validation_strategy_metadata_URI
-                );
+                self
+                    .emit(
+                        Event::ProposalValidationStrategyUpdated(
+                            ProposalValidationStrategyUpdated {
+                                proposal_validation_strategy: input
+                                    .proposal_validation_strategy
+                                    .clone(),
```

Comment:
not sure whether there is a way to avoid the clones here? 

Now eventstake a struct, you cannot pass a snapshot to one of the fields like we did before.

Reply:
yeah kinda sucks. I wonder why the `write` to storage function takes ownership of the data. Would be better if it were using a `@`...